src/iceberg/snapshot.h

```diff
@@ -0,0 +1,213 @@
+  /// \brief Number of duplicate files deleted (duplicates are files recorded more than
+  /// once in the manifest)
+  constexpr static std::string_view kDeletedDuplicatedFiles = "deleted-duplicate-files";
+  /// \brief Number of partitions with files added or removed in the snapshot
+  constexpr static std::string_view kChangedPartitionCountProp =
+      "changed-partition-count";
+
+  /// Other Fields, see https://iceberg.apache.org/spec/#other-fields
+
+  /// \brief The Write-Audit-Publish id of a staged snapshot
+  constexpr static std::string_view kWAPID = "wap.id";
+  /// \brief The Write-Audit-Publish id of a snapshot already been published
+  constexpr static std::string_view kPublishedWAPID = "published-wap-id";
+  /// \brief The original id of a cherry-picked snapshot
+  constexpr static std::string_view kSourceSnapshotID = "source-snapshot-id";
+  /// \brief Name of the engine that created the snapshot
+  constexpr static std::string_view kEngineName = "engine-name";
+  /// \brief Version of the engine that created the snapshot
+  constexpr static std::string_view kEngineVersion = "engine-version";
+};
+
+/// \brief Summarises the changes in the snapshot.
+class ICEBERG_EXPORT Summary : public iceberg::util::Formattable {
+ public:
+  /// \brief The operation field is used by some operations, like snapshot expiration, to
+  /// skip processing certain snapshots.
+  enum class Operation {
+    /// Only data files were added and no files were removed.
+    kAppend,
+    /// Data and delete files were added and removed without changing table data; i.e.
+    /// compaction, change the data file format, or relocating data files.
+    kReplace,
+    /// Data and delete files were added and removed in a logical overwrite operation.
+    kOverwrite,
+    /// Data files were removed and their contents logically deleted and/or delete files
+    /// were added to delete rows.
+    kDelete,
+  };
+  Summary() = default;
+  /// \brief Construct a summary with the given operation and properties.
+  Summary(Operation op, std::unordered_map<std::string, std::string> props);
+
+  /// \brief Get the operation type of the snapshot.
+  Operation operation() const;
+
+  /// \brief Get the additional properties of the snapshot.
+  const std::unordered_map<std::string, std::string>& properties() const;
+
+  std::string ToString() const override;
+
+ private:
+  /// The type of operation in the snapshot
+  Operation operation_{Operation::kAppend};
+  /// Other summary data.
+  std::unordered_map<std::string, std::string> additional_properties_;
+};
+
+/// \brief A snapshot of the data in a table at a point in time.
+///
+/// A snapshot consist of one or more file manifests, and the complete table contents is
+/// the union of all the data files in those manifests.
+///
+/// Snapshots are created by table operations.
+class ICEBERG_EXPORT Snapshot : public iceberg::util::Formattable {
```

Comment:
Is it much simpler to use a struct without those trivial getters and constructor for `Snapshot`? I did this to `TableMetadata` because it is a metadata-carrying object without any logic. We can also use [aggregate initialization](https://en.cppreference.com/w/cpp/language/aggregate_initialization) to deal with optional fields in a more flexible way. I think `Snapshot` and even `SortOrder` share the same attribute.

Reply:
changed.